core/src/main/scala/org/apache/spark/scheduler/BlacklistTracker.scala

```diff
@@ -0,0 +1,393 @@
+   */
+  val taskToFailureCountAndExpiryTime = HashMap[Int, (Int, Long)]()
+  def updateWithFailure(taskIndex: Int, failureExpiryTime: Long): Unit = {
+    val (prevFailureCount, prevFailureExpiryTime) =
+      taskToFailureCountAndExpiryTime.getOrElse(taskIndex, (0, -1L))
+    assert(failureExpiryTime >= prevFailureExpiryTime)
+    taskToFailureCountAndExpiryTime(taskIndex) = (prevFailureCount + 1, failureExpiryTime)
+  }
+  def numUniqueTasksWithFailures: Int = taskToFailureCountAndExpiryTime.size
+
+  override def toString(): String = {
+    s"numUniqueTasksWithFailures= $numUniqueTasksWithFailures; " +
+      s"tasksToFailureCount = $taskToFailureCountAndExpiryTime"
+  }
+}
+
+/**
+ * Tracks all failures for one executor (that have not passed the timeout).  Designed to efficiently
+ * remove failures that are older than the timeout, and query for the number of unique failed tasks.
+ * In general we actually expect this to be extremely small, since it won't contain more than the
+ * maximum number of task failures before an executor is failed (default 2).
+ */
+private[scheduler] final class ExecutorFailureList extends Logging {
+
+  private case class TaskId(stage: Int, stageAttempt: Int, taskIndex: Int)
+
+  /**
+   * All failures on this executor in successful task sets, sorted by time ascending.
+   */
+  private var failures = ArrayBuffer[(TaskId, Long)]()
+
+  def addFailures(
+      stage: Int,
+      stageAttempt: Int,
+      failuresInTaskSet: ExecutorFailuresInTaskSet): Unit = {
+    // The new failures may interleave with the old ones, so rebuild the failures in sorted order.
+    // This shouldn't be expensive because if there were a lot of failures, the executor would
+    // have been blacklisted.
+    if (failuresInTaskSet.taskToFailureCountAndExpiryTime.nonEmpty) {
+      failuresInTaskSet.taskToFailureCountAndExpiryTime.foreach { case (taskIdx, (_, time)) =>
+        failures += ((TaskId(stage, stageAttempt, taskIdx), time))
+      }
+      // sort by failure time, so we can quickly determine if any failure has gone past the timeout
+      failures = failures.sortBy(_._2)
+    }
+  }
+
+  def minExpiryTime: Long = failures.headOption.map(_._2).getOrElse(Long.MaxValue)
+
+  /**
+   * The number of unique tasks that failed on this executor.  Only counts failures within the
+   * timeout, and in successful tasksets.
+   */
+  def numUniqueTaskFailures: Int = failures.size
+
+  def isEmpty: Boolean = failures.isEmpty
+
+  def dropFailuresWithTimeoutBefore(dropBefore: Long): Unit = {
+    if (minExpiryTime < dropBefore) {
+      val minIndexToKeep = failures.indexWhere(_._2 >= dropBefore)
+      if (minIndexToKeep == -1) {
+        failures.clear()
+      } else {
+        failures = failures.drop(minIndexToKeep)
+      }
+    }
+  }
+
+  override def toString(): String = {
+    s"failures = $failures"
+  }
+}
+
+private final case class BlacklistedExecutor(node: String, expiryTime: Long)
```

Comment:
good point, moved this and `ExecutorFailureList` as well.  I'd prefer to keep it as a case class since there are uses like `executorIdToBlacklistStatus.filter(_._2.expiryTime < now).keys` and I think that is much clearer with the `expiryTime` in there.